yt/frontends/cf_radial/io.py

```diff
@@ -0,0 +1,67 @@
+"""
+CF-Radial-specific IO functions
+
+
+
+"""
+
+import numpy as np
+
+from yt.utilities.io_handler import BaseIOHandler
+
+
+class CFRadialIOHandler(BaseIOHandler):
+    _particle_reader = False
+    _dataset_type = "cf_radial"
+
+    def _read_particle_coords(self, chunks, ptf):
+        # This needs to *yield* a series of tuples of (ptype, (x, y, z)).
+        # chunks is a list of chunks, and ptf is a dict where the keys are
+        # ptypes and the values are lists of fields.
+        pass
+
+    def _read_particle_fields(self, chunks, ptf, selector):
+        # This gets called after the arrays have been allocated.  It needs to
+        # yield ((ptype, field), data) where data is the masked results of
+        # reading ptype, field and applying the selector to the data read in.
+        # Selector objects have a .select_points(x,y,z) that returns a mask, so
+        # you need to do your masking here.
+        pass
+
+    def _read_fluid_selection(self, chunks, selector, fields, size):
+        # This needs to allocate a set of arrays inside a dictionary, where the
+        # keys are the (ftype, fname) tuples and the values are arrays that
+        # have been masked using whatever selector method is appropriate.  The
+        # dict gets returned at the end and it should be flat, with selected
+        # data.  Note that if you're reading grid data, you might need to
+        # special-case a grid selector object.
+        # Also note that "chunks" is a generator for multiple chunks, each of
+        # which contains a list of grids. The returned numpy arrays should be
+        # in 64-bit float and contiguous along the z direction. Therefore, for
+        # a C-like input array with the dimension [x][y][z] or a
+        # Fortran-like input array with the dimension (z,y,x), a matrix
+        # transpose is required (e.g., using np_array.transpose() or
+        # np_array.swapaxes(0,2)).
+        chunks = list(chunks)
+
+        rv = {}
+        for field in fields:
+            rv[field] = np.empty(size, dtype="float64")
+
+        offset = 0
+
+        for field in fields:
+            for chunk in chunks:
+                for grid in chunk.objs:
+                    ds = self.ds._handle
+                    variable = ds.variables[field[1]]
+                    data = variable.values[0, ...].T
```

Comment:
the transposition may have undesirable perfs impact and could probably be avoided by writing the data in the appropriate order from the start.

Reply:
I don't follow this -- these netcdf files are what they are, we won't have control over the ordering they're written in.

Reply:
Honestly I'll admit I might be very confused about this. I _think_ transpositions are very rarely necessary because oftentimes they are used to compensate for arrays that are read in the wrong order ("F" VS "C"). This might be either wrong or irrelevant here, I'm sorry, never managed to properly get my head around this problem.

Reply:
no worries! I find it very confusing as well... I always end up with a lot of trial and error with netcdf index and coordinate ordering... I do think it's needed in this case though.

Reply:
I don't think this should be a blocker, I was overcorrecting for my lack of understanding of what numpy does with transposition. Let's just resolve this thread